x-pack/plugins/lens/public/indexpattern_datasource/operations/layer_helpers.ts

```diff
@@ -426,12 +446,28 @@ export function deleteColumn({
 }
 
 export function getColumnOrder(layer: IndexPatternLayer): string[] {
-  const [aggregations, metrics] = _.partition(
+  const [direct, referenceBased] = _.partition(
     Object.entries(layer.columns),
-    ([id, col]) => col.isBucketed
+    ([id, col]) => operationDefinitionMap[col.operationType].input !== 'fullReference'
   );
+  // If a reference has another reference as input, put it last in sort order
```

Comment:
I'm not sure this catches all possible cases for longer chains of references (e.g. what if b doesn't reference a directly, but it references c which in turn references a?), but as we won't have them in the first version, I think it's fine for now if we leave a comment to revisit once we get more complex here.

Reply:
Right, this logic only handles chains of references with up to 2 levels. I will add a comment about this. I was thinking that the most correct way is to do [topological sorting](https://en.wikipedia.org/wiki/Topological_sorting), but we don't need it yet.